plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java

```diff
@@ -180,6 +196,69 @@ public void onFailure(Exception e) {
+        if (!enableAutomaticDeployment || (mlModelManager.isModelDeployed(modelId) && workerNodes != null && workerNodes.length != 0)) {
+            // model deployment finished, or automatic deployment disallowed
+            executePredict(mlPredictionTaskRequest, wrappedListener, modelId);
+            return;
+        }
+        mlModelManager.getModel(modelId, ActionListener.wrap(mlModel -> {
+            // Obtain the deployment plan
+            String[] workerNodeIds = mlModel.getPlanningWorkerNodes();
+            MLDeployModelRequest deployModelRequest = MLDeployModelRequest
+                .builder()
+                .modelId(modelId)
+                .async(false)
+                .dispatchTask(false)
+                .modelNodeIds(workerNodeIds)
+                .build();
+            deployAndPredict(modelId, mlModel, deployModelRequest, mlPredictionTaskRequest, wrappedListener);
+        }, exception -> {
+            log.error("Failed to find the remote model in the model index " + modelId, exception);
+            wrappedListener.onFailure(exception);
+        }));
+    }
+
+    private void deployAndPredict(
+        String modelId,
+        MLModel mlModel,
+        MLDeployModelRequest deployModelRequest,
+        MLPredictionTaskRequest mlPredictionTaskRequest,
+        ActionListener<MLTaskResponse> wrappedListener
+    ) {
+        ActionListener<MLDeployModelResponse> deployModelActionListener = ActionListener.wrap(deployModelResponse -> {
+            // Deployment failed, existing
+            if (!deployModelResponse.getStatus().equals(MLTaskState.COMPLETED.name())) {
+                wrappedListener
+                    .onFailure(
+                        new IllegalArgumentException(
+                            "Deployment Failed for " + modelId + ", please check your remote model before predicting."
+                        )
+                    );
+                return;
+            }
+            String[] workerNodes = mlModelManager.getWorkerNodes(modelId, FunctionName.REMOTE, true);
+            if (workerNodes == null || workerNodes.length == 0) {
+                // The deployment is kicked off, but not finished yet in this node
+                mlModelManager.deployRemoteModelToLocal(modelId, mlModel, ActionListener.wrap(r -> {
+                    mlPredictionTaskRequest.setDispatchTask(false);
+                    executePredict(mlPredictionTaskRequest, wrappedListener, modelId);
+                }, exception -> {
+                    log.error("Failed to deploy model to local node" + modelId, exception);
+                    wrappedListener.onFailure(exception);
+                }));
+            } else {
+                executePredict(mlPredictionTaskRequest, wrappedListener, modelId);
+            }
+        }, wrappedListener::onFailure);
+        client.execute(MLDeployModelAction.INSTANCE, deployModelRequest, deployModelActionListener);
```

Comment:
If we have 10k requests comes same time, they may trigger 10k deploy request ? I think we discussed this issue before, but seems this PR doesn't solve it

Reply:
This is a good point. First of all we need to agree that it's the best cx experience for 10k requests to finish with correct results rather than rejecting any of them. With that said, we can limit the number of transport deploy call to be maximum 1 through a synchronizedHashSet tracking each modelId. For the local deploys, we cannot limit the number if they needs to happen before the Transport Layer Deployment finishes. This is the price to pay for invoking a cold model without deployment, and it's not that expensive because the local deploy only executes a few doze of code to setup the "Predictable". 

In the example of 10K concurrent requests, the first request will invoke the "Deploy API" in the transport layer. There won't be 9999 local deploy since we don't have that many threads available. Local deploy will exit in the middle if it finds the deployment from transport deployment is finished. So a small portion of 9999 will be local deploy, and then the transport layer deployment finishes, and all rests invokes becomes normal happen case.